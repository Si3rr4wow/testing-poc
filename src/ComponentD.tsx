import { useParams } from "react-router-dom"
import ComponentB from "./ComponentB"

const ComponentD: React.FC = () => {
  const { id } = useParams<{ id: string }>()

  return (
    <div className="component-wrapper">
      <code className="component-label">
        <h1>
          Component D
        </h1>
        <p>
          This component grabs a person's id from a url param
        </p>
        <p>
          This component has a child which finds a person given an id
        </p>
      </code>
      <p>
        The ID in the url is {id}
      </p>
      <ComponentB id={id}/>
    </div>
  )
}

export default ComponentD