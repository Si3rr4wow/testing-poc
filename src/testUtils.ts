export const getRandomString = (length = 8): string => 
  Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, length);
