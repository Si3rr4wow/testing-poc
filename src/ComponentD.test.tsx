import { act, render, RenderResult } from '@testing-library/react'
import ComponentD from './ComponentD'
import findPerson from './findPerson'
import { getRandomString } from './testUtils'
import { MemoryRouter, Route } from 'react-router-dom'
jest.mock('./findPerson', () => ({
  __esModule: true,
  default: jest.fn()
}))

const renderComponentD = async (id: string): Promise<RenderResult> => {
  let renderResult = {} as RenderResult;
  await act(async () => {
    renderResult = render(
      <MemoryRouter initialEntries={[`/person/${id}`]}>
        <Route path="/person/:id">
          <ComponentD/>
        </Route>
      </MemoryRouter>
    )
  })
  return renderResult
}

describe('ComponentD', () => {
  // Integration test of D and B
  it('has a child that fetches the person with the id I typed', async () => {
    const id = getRandomString()
    const name = getRandomString();
    (findPerson as jest.Mock).mockImplementation(() => Promise.resolve({ name }))
    await renderComponentD(id)
    expect(findPerson).toHaveBeenCalledWith(id)
  })
  // Integration test of D, B and A
  it('matches it`s snapshot', async () => {
    const id = 'Snapshot id'
    const name = 'Snapshot name';
    (findPerson as jest.Mock).mockImplementation(() => Promise.resolve({ name }))
    const { asFragment } = await renderComponentD(id)
    expect(asFragment()).toMatchSnapshot()
  })
})