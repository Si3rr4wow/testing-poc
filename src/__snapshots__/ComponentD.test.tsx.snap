// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentD matches it\`s snapshot 1`] = `
<DocumentFragment>
  <div
    class="component-wrapper"
  >
    <code
      class="component-label"
    >
      Component D
    </code>
    <p>
      The ID in the url is Snapshot id
    </p>
    <div
      class="component-wrapper"
    >
      <code
        class="component-label"
      >
        Component B
      </code>
      <p>
        finding person with id Snapshot id
      </p>
      <div
        class="component-wrapper"
      >
        <code
          class="component-label"
        >
          Component A
        </code>
        <p>
          The person's name is Snapshot name
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
